src/mame/ensoniq/esq5505.cpp

```diff
@@ -294,15 +308,160 @@ class esq5505_state : public driver_device
 	void cpu_space_map(address_map &map) ATTR_COLD;
 	void eps_cpu_space_map(address_map &map) ATTR_COLD;
 
+	void update_floppy_inputs();
+	void floppy_loaded(bool loaded);
+	void floppy_load(floppy_image_device *floppy);
+	void floppy_unload(floppy_image_device *floppy);
+	void cartridge_loaded(bool loaded);
+	void cartridge_load(ensoniq_vfx_cartridge *cart);
+	void cartridge_unload(ensoniq_vfx_cartridge *cart);
+
+	void update_docirq_to_maincpu();
+	void otis_irq(int irq);
+
 	uint16_t m_analog_values[8];
 };
 
+void esq5505_state::cartridge_loaded(bool loaded)
+{
+	LOG("Cartridge %s\n", loaded ? "Inserted" : "Ejected");
+	int state = loaded ? CLEAR_LINE : ASSERT_LINE;
+
+	// On VFX and later, DUART input bit 1 is 0 for cartridge present.
+	LOG("ip1 -> %d\n", state);
+	m_duart->ip1_w(state);
+}
+
+void esq5505_state::cartridge_load(ensoniq_vfx_cartridge *cart)
+{
+	(void) cart;
+	cartridge_loaded(true);
+}
+
+void esq5505_state::cartridge_unload(ensoniq_vfx_cartridge *cart)
+{
+	(void) cart;
+	cartridge_loaded(false);
+}
+
+void esq5505_state::floppy_drives(device_slot_interface &device)
+{
+	device.option_add_internal("35dd", FLOPPY_35_DD);
+}
+
 void esq5505_state::floppy_formats(format_registration &fr)
 {
 	fr.add_mfm_containers();
 	fr.add(FLOPPY_ESQIMG_FORMAT);
 }
 
+TIMER_CALLBACK_MEMBER(esq5505_state::floppy_motor_on)
+{
+	bool motor_on = param;
+	if (m_floppy_connector)
+	{
+		floppy_image_device *floppy = m_floppy_connector->get_device();
+		if (floppy)
+		{
+			floppy->mon_w(motor_on ? CLEAR_LINE : ASSERT_LINE);  // active low
+			m_panel->set_floppy_active(motor_on);
+		}
+	}
+}
+
+TIMER_CALLBACK_MEMBER(esq5505_state::floppy_dskchg_reset)
+{
+	m_floppy_dskchg = !m_floppy_is_loaded;
+	LOG("Resetting floppy_dskchg -> %s\n", m_floppy_dskchg ? "true" : "false");
+	update_docirq_to_maincpu();
+}
+
+void esq5505_state::update_floppy_inputs()
+{
+	// update the "Disk Ready" input
+	int state = (m_floppy_is_active && m_floppy_is_loaded) ? ASSERT_LINE : CLEAR_LINE;
+#if VERBOSE
+	static int prev_state = 0;
```

Comment:
I would use `LOGMASKED` for this rather than #if-ing it. #if'd-out code tends to rot. Should be plenty of examples to choose from when it comes to `LOGMASKED`, the only thing to know is that `LOG(...)` is effectively an alias for `LOGMASKED(LOG_GENERAL, ...)`, so user-defined verbose masks should start at `1U << 1`.

Reply:
it's not about warnings but about having and updating a variable (`prev_state`) that is only there to limit logging to when the state actually changes. Neither the variable, nor the code updating it, are needed other than when logging is enabled.

Someone who reads the code can so much more clearly see that intent if the declaration of the variable and the code to updating it as well as the logging that uses it are all together inside a `#ifdef` guard. 

An intelligent compiler would hopefully see that, when `VERBOSE` is `0`, the variable is only ever written to and never read so that it can effectively be elided, but code is written not just for the compiler but also for programmers in the future who are trying to understand what is going on.